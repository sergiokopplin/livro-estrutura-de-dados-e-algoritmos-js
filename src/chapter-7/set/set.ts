export class Set {
  items: any;

  constructor() {
    this.items = {};
  }

  has(element: any): boolean {
    return Object.prototype.hasOwnProperty.call(this.items, element);
  }

  add(element: any): boolean {
    if (!this.has(element)) {
      this.items[element] = element;

      return true;
    }

    return false;
  }

  delete(element: any): boolean {
    if (this.has(element)) {
      delete this.items[element];
      return true;
    }

    return false;
  }

  clear(): void {
    this.items = {};
  }

  size(): number {
    return Object.keys(this.items).length;
  }

  values(): any[] {
    return Object.values(this.items);
  }
}
