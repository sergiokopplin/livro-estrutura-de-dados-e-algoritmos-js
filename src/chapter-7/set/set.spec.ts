import { Set } from './set';

describe('Set', () => {
  test('Should construct correctly', () => {
    const set = new Set();

    expect(set).toMatchSnapshot();
  });

  test('Should work properly', () => {
    const set = new Set();

    set.add(1);

    expect(set).toMatchSnapshot();
    expect(set.values()).toEqual([1]);
    expect(set.has(1)).toBe(true);
    expect(set.size()).toBe(1);

    set.add(2);

    expect(set).toMatchSnapshot();
    expect(set.values()).toEqual([1, 2]);
    expect(set.has(2)).toBe(true);
    expect(set.size()).toBe(2);

    set.delete(1);

    expect(set).toMatchSnapshot();
    expect(set.values()).toEqual([2]);

    set.delete(2);

    expect(set).toMatchSnapshot();
    expect(set.values()).toEqual([]);

    set.add(1);
    expect(set.values()).toEqual([1]);
    set.clear();
    expect(set.values()).toEqual([]);
  });
});
