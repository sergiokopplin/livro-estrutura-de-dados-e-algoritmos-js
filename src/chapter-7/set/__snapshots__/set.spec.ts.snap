// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set Should construct correctly 1`] = `
Set {
  "items": {},
}
`;

exports[`Set Should work properly 1`] = `
Set {
  "items": {
    "1": 1,
  },
}
`;

exports[`Set Should work properly 2`] = `
Set {
  "items": {
    "1": 1,
    "2": 2,
  },
}
`;

exports[`Set Should work properly 3`] = `
Set {
  "items": {
    "2": 2,
  },
}
`;

exports[`Set Should work properly 4`] = `
Set {
  "items": {},
}
`;
